pallets/author-inherent/src/lib.rs

```diff
@@ -81,51 +79,54 @@ pub mod pallet {
 		CannotBeAuthor,
 	}
 
-
 	/// Author of current block.
 	#[pallet::storage]
 	pub type Author<T: Config> = StorageValue<_, T::AccountId, OptionQuery>;
 
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
 		fn on_initialize(_: T::BlockNumber) -> Weight {
+			// Start by clearing out the previous block's author
 			<Author<T>>::kill();
+
+			// Now extract the author from the digest
+			let digest = <frame_system::Pallet<T>>::digest();
+
+			let pre_runtime_digests = digest.logs.iter().filter_map(|d| d.as_pre_runtime());
+			Self::find_author(pre_runtime_digests).map(|author_account|{
+				// Store the author so we can confirm eligibility after the inherents have executed
+				<Author<T>>::put(&author_account);
+
+				//TODO, should we reuse the same trait that Pallet Authorship uses?
+				// Notify any other pallets that are listening (eg rewards) about the author
+				T::EventHandler::note_author(author_account);
```

Comment:
The note_author method should return a Weight

Reply:
That's a good point. This trait was based off of https://crates.parity.io/pallet_authorship/trait.EventHandler.html which also does not return a weight.

As the comment says, I was also thinking of switching to using their trait.

Reply:
I've made a proposal to Parity in https://github.com/paritytech/substrate/issues/10437, and a followup nimbus ticket in https://github.com/PureStake/nimbus/issues/29.

I prefer not to block this already-huge PR on this since the same issue existed before the PR. But I do totally agree with your point.